packages/paima-sdk/paima-events/src/app-events.ts

```diff
@@ -0,0 +1,200 @@
+      // case the signature/topicHash needs to be computed again, which can't be
+      // done from the path (since you don't know which non indexed fields go in
+      // between each indexed field).
+      definition: event,
+      // we add this to avoid having to re-compute it all the time.
+      topicHash: toTopicHash(event),
+    });
+  }
+  return result as any;
+};
+
+// create payload for the stf from an object.
+export function createEventForStf<T extends ReturnType<typeof genEvent>>(
+  address: `0x${string}`,
+  event: T,
+  fields: KeypairToObj<T['fields']>
+): {
+  address: `0x${string}`;
+  data: { name: string; fields: KeypairToObj<T['fields']>; topic: string };
+} {
+  return { address, data: { name: event.name, fields, topic: toTopicHash(event) } };
+}
+
+// sum type for all the overloads of a fixed event name
+type UnionForOverloadedEvents<
+  T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+  N extends string,
+> = T extends (infer U)[]
+  ? U extends { name: N; fields: infer V }
+    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      V extends { name: string; type: any }[]
+      ? KeypairToObj<V>
+      : never
+    : never
+  : never;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type KeypairToObj<T extends { name: string; type: any }[]> = {
+  [K in T[number] as K['name']]: Static<K['type']>;
+};
+
+type EmitFunction<
+  T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+  Name extends string,
+> = {
+  emit: (name: Name, address: `0x${string}`, fields: UnionForOverloadedEvents<T, Name>) => void;
+};
+
+type EventQueue<T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>> = EmitFunction<
+  T,
+  T[number]['name']
+> & {
+  get: () => {
+    address: `0x${string}`;
+    data: { name: string; fields: KeypairToObj<T[number]['fields']>; topic: string };
+  }[];
+};
+
+// Typed stateful builder that handles overloaded topics. There is a performance
+// cost of using this, since it dynamically finds the matching overloaded topic
+// from the provided fields, but this way there is no need to use indexing to
+// provide the right event type. If this is a concern then the other helper
+// function can be used to explicitly provide the event type.
+//
+// This is supposed meant to be used like this (pseudo-code):
+//
+// const eventsDefinitions = [
+//   genEvent({ name: 'A', fields: fields1 }),
+//   genEvent({ name: 'B', fields: fields2 }),
+//   genEvent({ name: 'B', fields: fields3 }),
+// ] as const;
+//
+// const queueFactory = eventQueueFactory(eventsDefinitions);
+// const queue = queueFactory.empty();
+//
+// queue.emit('A', data1);
+// queue.emit('B', data2);
+// queue.emit('B', data3);
+//
+// return { events: queue, stateTransitions: {...}};
+//
+// Where data2 and data3 may be different types (for the two overloaded
+// definitions), and the return type is that of the stf.
+export function eventQueueFactory<
```

Comment:
The api on this may be slightly better (so it's currently optional), but I think at least this may be a starting point

Currently I've been using this like this:

By previously initializing the object at the beginning of the stf function. Basically something like this:

```ts
export default async function (
  inputData: SubmittedChainData,
  blockHeader: BlockHeader,
  randomnessGenerator: Prando,
  dbConn: Pool
): Promise<{
  stateTransitions: SQLUpdate[];
  events: any;
}> {
  const events = eventQueueFactory.empty();

  const user = inputData.userAddress.toLowerCase();
  const parsed = parse(inputData.inputData);

  switch (parsed.input) {
    case 'command1':
      handleCommand1(events, user, parsed.tag, blockHeader.blockHeight);
      break;
    case 'input2':
      handleCommand2(events, user, blockHeader.blockHeight);
      break;
  }

  return { stateTransitions: [], events: events.get() };
}

function handleCommand1(
  eventQueue: ReturnType<typeof eventQueueFactory.empty>,
  user: string,
  tag: string,
  block: number
) {
  eventQueue.emit('StfEvent', user as any, { tag: 'command1', data: 'test', extraData: tag });

  eventQueue.emit('StfEvent', user as any, {
    tag: 'command2_1',
    data: 'test',
    extraData: 'extra',
    block: block,
  });
}
```

I think my preferred API would to pass an instance of this queue as an argument to the state transitition function, because I think having to return an object of arrays looks weird. The problem is that this currently can't be typed since the app events are loaded at runtime.